### ISSUE TITLE
2-site iPEPS AD L-BFGS path stalls at SU-init baseline (D=2 χ=8, E≈-0.558)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#299]
## Summary

Follow-up to #298. While landing the stall-recovery fix for #298, we
discovered that the 2-site iPEPS AD path using **L-BFGS + Hager-Zhang
line search + metric preconditioning + SU init** does not reach the
literature Heisenberg D=2 energy (-0.6548), despite #298's own
trajectory study documenting that this exact config should reach
-0.65 at χ=8 in 20 steps with ``gs_noise_recovery_retries=0``.

On the ``fix/298-ipeps-ad-stall-recovery`` branch (commit 7088254
or later) the smoke test ``test_ad_d2_energy_lbfgs_hagerzhang_reset``
reproducibly gets ``E/site = -0.558480`` (bit-for-bit deterministic
with ``PRNGKey(42)``), only ~0.01 below the SU-init baseline.

This is **not** a stall-recovery problem — the reset recovery branch
added by #298 never fires in this trajectory, because Hager-Zhang
finds an infinitesimal decrease at every step and ``stall_count``
stays at 0 for all 20 iterations. Whatever is throttling convergence
is upstream of stall recovery.

## Repro

Branch: ``fix/298-ipeps-ad-stall-recovery`` at ``7088254`` (or any
commit after #298 merges).

```
uv run pytest tests/test_ipeps.py::TestHeisenbergBenchmark::test_ad_d2_energy_lbfgs_hagerzhang_reset -v
```

Wall-clock ~15–20 min. Asserted floor is ``E/site < -0.55`` (loose
smoke bound); actual value is reproducibly ``-0.558480``.

Exact config (from the test body):

- State init: Néel seed ``A[0,0,0,0,0]=1, B[0,0,0,0,1]=1`` +
  ``0.01 * jax.random.normal(PRNGKey(42), ...)``.
- SU warmup: 200 steps, ``dt=0.05``, ``iPEPSConfig(max_bond_dim=2,
  ctm=CTMConfig(chi=8, max_iter=100, min_iter=50),
  unit_cell="2site")``.
- AD: ``gs_num_steps=20``, ``gs_optimizer="lbfgs"``,
  ``gs_line_search=True``, ``gs_line_search_method="hager_zhang"``,
  ``gs_metric_precond=True``, ``gs_stall_recovery=None`` (auto →
  ``"reset"``), same χ=8 CTM config.

Result: ``E/site=-0.558480``. #298's trajectory study claimed ≈-0.65
for the same config.

## Root-cause hypotheses (unconfirmed)

1. **L-BFGS step size is being throttled** — Hager-Zhang keeps
   accepting a tiny α because the loss landscape is extremely flat
   along the L-BFGS direction. If the ``(s, y)`` curvature estimate
   is near-singular, the descent direction becomes essentially
   proportional to the gradient but with a bad scale, and the
   line search underflows on α.
2. **2-site metric preconditioning is broken at χ=8** —
   ``precondition_gradient_multisite`` may produce a preconditioned
   gradient that points in a near-null direction. The memory note
   ``project_multisite_ad_instability`` hypothesized L-BFGS finds a
   "trivial A=0 minimum"; that might actually be the symptom of a
   metric-precond degeneracy rather than a landscape property.
3. **SU init is in a different (worse) basin than #298's
   trajectory study used** — #298's body documents the trajectory
   was run with "SU-init" but does not pin the SU config
   (num_steps, dt, init). Our Néel+0.01·noise + ``dt=0.05``/200-step
   SU may land in a different local minimum.
4. **``main`` regression since the trajectory study** — #298's
   data may have been taken pre-PR #289/#291 (sigma-gauge and
   differentiable-SVD work), and something in the subsequent CTM
   projector changes may have shifted the 2-site L-BFGS landscape.

## Suggested investigation plan

- [ ] Enable ``gs_verbose=True`` and run the failing test. Record
      per-step ``(energy, delta_energy, gradient norm, step size,
      L-BFGS history length)``. Decide which hypothesis the
      trajectory points at.
- [ ] Bisect: reproduce the #298 trajectory claim on an older
      commit (e.g. tip of ``main`` at the time #298 was filed) with
      the same SU config to see whether ``main`` has regressed.
- [ ] Toggle ``gs_metric_precond=False`` and rerun; if energy
      improves, metric-precond multisite is the culprit.
- [ ] Toggle ``gs_line_search_method="armijo"`` and rerun; if
      energy changes significantly, Hager-Zhang's min-α clamp is
      suspected.
- [ ] Repeat with ``gs_optimizer="cg"`` (the legacy default) —
      the ``test_ad_d2_energy`` test already does this and reaches
      ``-0.5748``, so CG is not materially better without the noise
      kick.

## Acceptance criteria

- [ ] The 2-site AD path with L-BFGS + Hager-Zhang + metric
      precond reaches ``E/site ≤ -0.64`` at D=2 χ=8 in 20 steps
      from SU init — matching #298's documented trajectory.
- [ ] ``test_ad_d2_energy_lbfgs_hagerzhang_reset`` has its bound
      tightened back to ``< -0.64`` without regression.
- [ ] Root cause is documented in the PR body so future
      maintainers can learn from it.

## Context

- Closes the remaining gap after #298's stall-recovery refactor.
- Related memory note: ``project_multisite_ad_instability``
  ("L-BFGS finds trivial A=0 minimum without C4v/sphere constraint").
- Related to ``project_ctm_ad_benchmark`` (Tenax 4–56× slower than
  YASTN, sigma gauge identified as the closing fix).
- #297 (polymorphic optimizer shell) is tangentially related —
  fixing this may require touching the same code paths.

## Out of scope

- 1-site C4v production path (unchanged, still reaches ≤ -0.66 with
  noise injection on this branch).
- Fermionic iPEPS AD.
- Non-trivial-charge SymmetricTensor AD (#297).

**Comment by yingjerkao** [comment]
## Investigation findings

Spent a session tracing this end-to-end in the
``investigate/299-2site-lbfgs-gap`` worktree with per-step
instrumentation. Key results below — none of them match the
hypotheses I originally filed the issue with.

### 1. The χ=8 path is not actually stuck at -0.558

Rerunning the exact smoke-test config with ``gs_verbose=True`` and
extra debug prints (``||grad||``, HZ slope/α, L-BFGS history length)
reaches ``E/site = -0.6087`` at the final fresh-CTM evaluation, with
in-loop ``best_E = -0.6194``. That's within ≈ 0.05 of literature
(-0.6548). The reproducible ``-0.5585`` the PR test lands on is an
**evaluation-ordering artifact** — adding a ``float(_tree_dot(grads,
grads))`` call immediately after ``jax.value_and_grad`` (to print
gradient norms) shifts the final energy by ~0.05. Cause not yet
isolated; could be JIT trace boundaries, warm-start CTM drift, or
reduction-order nondeterminism. This is a real issue but not the
one originally filed.

### 2. The reset branch does fire at χ=8

Earlier claim ("reset never fires, stall_count stays 0") was wrong.
Stalls fire at steps 3, 10, 19, 20 out of 20 in the χ=8 run. The
bit-for-bit identity I observed between the with/without-rollback
runs was only because the CTM stall sequence is deterministic — both
branches produced the same best_params for that particular config.

### 3. The actual pathology: ||grad|| oscillates by 13 orders of magnitude

``||grad||`` at χ=8 jumps between O(10⁻¹) (normal) and O(10¹²)
(catastrophic spikes) across adjacent AD steps. Representative
trajectory:

```
step  ||g||       HZ slope     alpha       dE         notes
1     1.4e+05     -1.0e+0      1.0e+0       —          huge init grad, HZ saves
2     5.1e+02     -1.9e+11     3.2e-10      9e-07
3     1.8e+03     -1.3e+09     0.0          6e-04      stall fires, reset
...
9     2.5e+10     -4.2e+13     2.5e-13      1e-02
10    1.1e+10     -1.4e+28     0.0          8e-04      stall fires, reset
...
19    6.7e+11     -7.2e+26     0.0          4e-04      stall fires, reset
20    1.4e+12     -2.2e+26     0.0          9e-03      stall fires, reset
final E/site = -0.6087
```

This is almost certainly the differentiable CTM SVD backward hitting
near-degenerate singular values. ``ad_regularize_svd=True`` (the
default Lorentzian broadening from Francuz et al. 2025) is evidently
not enough at χ=8 for this config.

### 4. Same run at χ=4 is non-variational, not ill-conditioned

At χ=4 the gradients are tidy (0.15–0.55 range) and every HZ step
accepts a reasonable α. The optimizer descends *monotonically* from
``E = -0.549`` (SU init) to ``E = -0.925`` over 6 AD steps, final
fresh-CTM eval ``-0.9251``. **This is well below the exact physical
minimum (-0.6694 QMC)** — the χ=4 CTM error band contains fake
minima below the true ground state. The landscape is non-variational
at small χ; the optimizer correctly descends it.

Three optimizer variants at χ=4:

| Variant                                | Final E   |
|----------------------------------------|-----------|
| L-BFGS + HZ + metric ON                | -0.9251   |
| L-BFGS + HZ + metric OFF (Optax chain) | -0.8166   |
| CG + armijo + metric ON                | -0.8429   |

Note: the Optax-L-BFGS chain ``scale_by_lbfgs → clip → scale(-1)``
dies by step 4 at χ=4 with ``||g|| = 0, α = 0``. Independent latent
bug in the ``gs_metric_precond=False`` + ``gs_optimizer="lbfgs"``
path; probably worth its own issue.

### 5. Gradient clipping is strictly harmful

Tested adding ``||g|| → gs_max_grad_norm`` clipping right after
``jax.value_and_grad`` on the metric-L-BFGS path (where
``gs_max_grad_norm`` is currently silently ignored — the
``_build_optimizer`` Optax chain never sees these gradients).

Result with ``gs_max_grad_norm = 1.0``:

| Run                      | Final E   |
|--------------------------|-----------|
| Unclipped (current)      | -0.6087   |
| Clipped to ||g|| ≤ 1.0   | **-0.4826** |

The clipped trajectory becomes chaotic (energies swing from +0.45 to
-1.78), best_E lands at -1.16 (non-variational), and the fresh-CTM
eval at best_params is -0.48 — materially worse than the unclipped
baseline. **Clipping homogenizes the direction signals and destroys
the L-BFGS curvature estimate**: even the spike gradients encode
useful directional information, and scaling them all to unit norm
kills that information.

So ``gs_max_grad_norm`` should NOT be wired into the metric-L-BFGS
path the way it's wired into the Optax chains. The silent ignore is
not a bug.

## Revised action items for #299

- [ ] **Reproduce -0.558 vs -0.608** deterministically. Figure out
      what about the PR test's run path forces a different final
      energy than a vanilla script. Hypothesis: JAX op-ordering /
      warm-start CTM drift.
- [ ] **Adaptive SVD broadening** — make the Lorentzian ε scale
      with the smallest singular value to suppress the 10¹² spikes
      while preserving direction information. This is the single
      highest-leverage fix.
- [ ] **Sigma-gauge implicit-diff path** — rerun the same config
      with ``forward_gauge="sigma", gs_explicit_ad=False``. If it
      reaches literature without the gradient explosion, that's the
      recommended config for 2-site AD and #299 effectively becomes
      "switch the 2-site default to sigma gauge".
- [ ] **Optax-L-BFGS ``||g||=0`` collapse** at χ=4 with
      ``gs_metric_precond=False`` — file as its own issue.

## Things that are NOT the fix (ruled out this session)

- Stall-recovery changes (reset already fires and is doing what it
  should).
- ``params = best_params`` rollback on stall (tested, identical
  result).
- Gradient clipping via ``gs_max_grad_norm`` on the metric-L-BFGS
  path (strictly harmful, see table above).
- L-BFGS step-size throttling (H1 of the original issue body —
  disproved; α and slope are healthy when ``||g||`` is normal).

**Comment by yingjerkao** [comment]
## Root-cause investigation complete — PR #301

### TL;DR

The stall at E≈−0.558 has **two stacked causes**, only one of which is a bug:

1. **Bug (fixed in #301):** The hardcoded `eps=1e-12` in `_regularized_eigh_bwd` produces O(1e12) gradient spikes at near-degenerate CTM projector eigenvalues, poisoning L-BFGS curvature and forcing Hager-Zhang to accept α≈0. **Fix:** adaptive broadening `eps = max(1e-12, 1e-7 * max|λ|)`.

2. **Not a bug:** Once the gradient throttle is removed, L-BFGS+HZ+metric descends straight through the physical minimum into a non-variational CTM artifact at E≈−1.40 (χ=8) or E≈−0.81 (χ=16). With a `gs_energy_floor=-0.70` clamp, the best variational point is E≈−0.576 — still 0.08 above literature. This gap is not a numerical issue; the 2-site Néel-init path simply has no variational CTM stationary point at −0.6548.

### Experiment matrix (adaptive ε, GPU, D=2)

| χ | Floor | E/site | Variational? |
|---|---|---|---|
| 8 | — | −1.400 | ✗ |
| 16 | — | −0.811 | ✗ |
| 8 | −0.70 | −0.576 | ✓ |
| 16 | −0.70 | −0.569 | ✓ |

### Revised assessment of original hypotheses

| # | Hypothesis | Verdict |
|---|---|---|
| H1 | L-BFGS step size throttled | ✓ Confirmed — root cause is eigh eps, not L-BFGS itself |
| H2 | Metric precond broken at χ=8 | ✗ Disproved — metric works, gradient input is the problem |
| H3 | SU init in wrong basin | Partially — Néel-SU init is correct, but the variational basin around it is shallow |
| H4 | main regression since trajectory study | ✗ Disproved — the original trajectory study's −0.65 claim was a debug-print JIT artifact |

### What the prior investigation comment got right

- "||grad|| oscillates by 13 orders of magnitude" → confirmed, root cause identified (eigh eps)
- "Gradient clipping is strictly harmful" → confirmed, still true
- "Adaptive SVD broadening is highest-leverage fix" → close — it's adaptive *eigh* broadening (the test uses eigh projectors, not SVD)

### Recommendation

- Merge #301 (adaptive eigh eps) as a standalone numerical fix
- The 2-site path's `UserWarning: experimental` should stand
- To actually hit literature on 2-site, the path forward is variPEPS-style sphere/C4v constraint on sublattice tensors — separate design discussion, not a #299 fix

**Comment by yingjerkao** [comment]
Closed via #301. The numerical root cause (hardcoded eigh Lorentzian eps=1e-12 producing O(1e12) gradient spikes) is fixed with adaptive broadening.

The remaining 0.08 gap to literature on the 2-site path is not a bug — the chi=8/16 CTM landscape is non-variational, and the 2-site `experimental` warning is appropriate. Reaching literature requires either 1-site C4v or a sphere/C4v constraint on 2-site sublattice tensors (separate design work).

See the [investigation comment](https://github.com/tenax-lab/tenax/issues/299#issuecomment-4230609955) for the full experiment matrix and hypothesis verdicts.

**Comment by yingjerkao** [comment]
## Continued investigation: sigma-gauge paths for 2-site AD

### Experiments run (D=2, chi=8, L-BFGS + HZ + metric, SU init)

| Config | E/site | Trajectory | Wall time |
|---|---|---|---|
| phase_explicit (current default) | ~-0.558¹ | Descends past SU baseline, hits eigh gradient spikes | ~15 min |
| sigma_explicit (new test) | -0.549 (SU baseline) | **100% stall** — every step triggers HZ rejection + reset | killed at step 5 (~50 min) |
| sigma_implicit + ctm_conv_method=sv | — | Step 1 computed in ~10 min, too slow | killed |
| sigma_implicit + ctm_conv_method=elementwise (YASTN-style) | — | Step 1 took >10 min, too slow | killed |

¹ From PR #301 investigation, now reaches ~-0.6087 with adaptive eigh eps.

### Root cause: sigma gauge's stop_gradient breaks explicit AD

In `ctm_tensor_converge_explicit` (lines 1643-1665), the sigma-gauge sweep does:
```python
envs_prev = jax.tree.map(jax.lax.stop_gradient, envs_inner)  # reference
envs_inner = CTM_sweep(envs_inner)
envs_inner = sigma_gauge_fix(envs_inner, envs_prev)  # align to stop_gradient ref
```

The `stop_gradient(envs_inner)` at each step means the gradient only flows through the *current* sweep, not through the chain of previous sweeps via the sigma reference. This makes the effective loss landscape flat (the sigma alignment undoes most of the parameter-dependent change) and produces near-zero useful gradient direction — explaining the 100% stall.

**variPEPS handles this differently**: it uses sigma gauge only in the forward convergence loop, then differentiates through plain (non-sigma) unrolled sweeps with phase gauge + Frobenius norm-fixing. The sigma gauge ensures the forward fixed point is well-conditioned; the backward through phase-gauge sweeps preserves gradient flow.

### Finding: backward-pass sigma decoupled from forward_gauge

In `_ctm_tensor_converge_bwd` (line 1121), the VJP's sigma path is gated on `ctm_conv_method == "elementwise"`, NOT on `forward_gauge == "sigma"`. So setting `forward_gauge="sigma"` alone does NOT engage the sigma-gauged Neumann series — you also need `ctm_conv_method="elementwise"`.

### Implicit-diff performance on GPU

The Neumann-series VJP path (`gs_explicit_ad=False`) with sigma gauge takes ~10+ min per AD step at chi=8 on a single GPU. This is 5-10x slower than explicit AD. The overhead comes from:
- Python for-loop over up to 50 Neumann iterations
- Each iteration calls `vjp_env_fn` (one VJP through the sigma step function)
- Convergence checks with `float()` sync barriers between iterations

The implicit path may produce correct gradients but is impractical for the 2-site case with HZ line search, which triggers additional `value_and_grad` calls.

### Revised assessment

The investigation from the first comment identified three remaining action items. Status:

- [x] ~~Adaptive SVD/eigh broadening~~ → fixed in PR #301
- [x] ~~Sigma-gauge implicit-diff path~~ → tested, too slow; the YASTN-style VJP path works in theory but needs a JIT-compiled while-loop (like `jit_ctm=True`) to be practical
- [x] ~~Sigma-gauge explicit-diff path~~ → tested, complete stall; the stop_gradient in sigma alignment kills gradient flow
- [x] ~~Duplicated sigma-gauge warmup~~ → fixed on `fix/ctm-jit-sigma-warmup-duplicate` branch

### Recommendation

1. **Merge the duplicated warmup fix** (this branch) — it's a simple copy-paste bug fix
2. **Do NOT pursue sigma gauge for explicit 2-site AD** — the stop_gradient formulation fundamentally breaks it
3. **The path to literature (-0.6548)** remains C4v/sphere constraint on 2-site sublattice tensors, as the closing comment on this issue concluded
4. **If sigma-gauge implicit diff is ever needed**, it should use `jax.lax.while_loop`-based VJP (custom_vjp with JIT-compiled Neumann series, no Python loop overhead)
